### PR TITLE
MFT: introduce the decoder for the new data format.

### PR DESCRIPTION
**Pull request by bovulpes**
I add also a device to read the piped decoded digits and write them into a file in ROOT format. This is the full chain to check the encoding / decoding:

o2-sim -j 4 -m MFT -e TGeant3 -n 10 -g boxgen --configKeyValues 'BoxGun.pdg=13; BoxGun.eta[0]=-3.8; BoxGun.eta[1]=-2.1; BoxGun.prange[0]=5.0; BoxGun.prange[1]=10.; BoxGun.number=10'

o2-sim-digitizer-workflow -b 

o2-mft-digi2raw -b

o2-raw-file-reader-workflow -b --nocheck-missing-stop --max-tf 100 --input-conf MFTraw.cfg | o2-itsmft-stf-decoder-workflow -b --mft --digits --dict-file ITSdictionary.bin | o2-mft-digit-workflow -b

and at the end compare the digits from the two files:

mftdigits.root

and

mftdigitsdecode.root

### CODE REVIEW

File: Common/Utils/src/ConfigurableParam.cxx [hunk 1]
```diff
@@ -448,10 +448,12 @@ void ConfigurableParam::updateFromString(std::string const& configString)
 
   // Take a vector of strings with elements of form a=b, and
   // return a vector of pairs with each pair of form <a, b>
+  printf("BV ConfigurableParam::updateFromString configString %s \n", configString.data());
```
Comment (shahor02) [inline Common/Utils/src/ConfigurableParam.cxx:451]: this seems to be a private debug message, to be removed

File: Common/Utils/src/ConfigurableParam.cxx [hunk 2]
```diff
@@ -448,10 +448,12 @@ void ConfigurableParam::updateFromString(std::string const& configString)
...
   auto toKeyValPairs = [](std::vector<std::string>& tokens) {
     std::vector<std::pair<std::string, std::string>> pairs;
 
     for (auto& token : tokens) {
+      printf("BV ConfigurableParam::updateFromString token %s \n", token.data());
```
Comment (shahor02) [inline Common/Utils/src/ConfigurableParam.cxx:456]: idem

File: Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx [hunk 3]
```diff
@@ -103,7 +103,13 @@ void STFDecoder<Mapping>::run(ProcessingContext& pc)
...
   mTimer.Start(false);
   mDecoder->startNewTF(pc.inputs());
-  auto orig = o2::header::gDataOriginITS;
+  auto detID = Mapping::getDetID();
+  o2::header::DataOrigin orig;
+  if (detID == o2::detectors::DetID::ITS) {
+    orig = o2::header::gDataOriginITS;
+  } else {
+    orig = o2::header::gDataOriginMFT;
+  }
```
Comment (shahor02) [inline Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx:112]: You can use directly [``detID.getDataOrigin()``](https://github.com/AliceO2Group/AliceO2/blob/11f6a7db21bf3cad71a06a4ffbfecf3027a13ed2/DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h#L90)

File: Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx [hunk 4]
```diff
@@ -124,13 +130,21 @@ void STFDecoder<Mapping>::run(ProcessingContext& pc)
 
   if (mDoDigits) {
     pc.outputs().snapshot(Output{orig, "DIGITS", 0, Lifetime::Timeframe}, digVec);
-    pc.outputs().snapshot(Output{orig, "ITSDigitROF", 0, Lifetime::Timeframe}, digROFVec);
+    if (detID == o2::detectors::DetID::ITS) {
+      pc.outputs().snapshot(Output{orig, "ITSDigitROF", 0, Lifetime::Timeframe}, digROFVec);
+    } else {
+      pc.outputs().snapshot(Output{orig, "MFTDigitROF", 0, Lifetime::Timeframe}, digROFVec);
+    }
```
Comment (shahor02) [inline Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx:137]: could you instead change all ``ITSDigitROF`` and ``MFTDigitROF`` to ``DigitROF``, the same for ``ClusterROF``? The  DataDescription always comes with DataOrigin, no need for extra naming scheme which make the code more complex

File: Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx [hunk 5]
```diff
@@ -124,13 +130,21 @@ void STFDecoder<Mapping>::run(ProcessingContext& pc)
...
   }
   if (mDoClusters) {                                                                  // we are not obliged to create vectors which are not requested, but other devices might not know the options of this one
     pc.outputs().snapshot(Output{orig, "CLUSTERS", 0, Lifetime::Timeframe}, clusVec); // DUMMY!!!
     pc.outputs().snapshot(Output{orig, "COMPCLUSTERS", 0, Lifetime::Timeframe}, clusCompVec);
     pc.outputs().snapshot(Output{orig, "PATTERNS", 0, Lifetime::Timeframe}, clusPattVec);
-    pc.outputs().snapshot(Output{orig, "ITSClusterROF", 0, Lifetime::Timeframe}, clusROFVec);
+    if (detID == o2::detectors::DetID::ITS) {
```
Comment (shahor02) [inline Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx:143]: here too

File: Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx [hunk 6]
```diff
@@ -213,7 +227,7 @@ DataProcessorSpec getSTFDecoderMFTSpec(bool doClusters, bool doPatterns, bool do
   std::vector<OutputSpec> outputs;
   auto orig = o2::header::gDataOriginMFT;
   if (doDigits) {
-    outputs.emplace_back(orig, "DIGMFT", 0, Lifetime::Timeframe);
+    outputs.emplace_back(orig, "DIGITS", 0, Lifetime::Timeframe);
     outputs.emplace_back(orig, "MFTDigitROF", 0, Lifetime::Timeframe);
```
Comment (shahor02) [inline Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx:231]: here too

File: Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx (hunk 3 above)
Comment (bovulpes) [reply Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx:112]: Hi Ruben,

I don't understand, detID is an integer.

Comment (shahor02) [reply Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx:112]: I meant ``DetID::getDataOrigin( Mapping::getDetID() )``. But I see that I've already implemented direct alias ``Mapping::getOrigin()``

**Comment by bovulpes** [comment]
PR #3715 is breaking the decoder for the MFT simulations. I am looking into this.

**Comment by bovulpes** [comment]
Hi @shahor02  I am pushing the reviewed code to work on top of your last PR #3715.

**Comment by shahor02** [comment]
@bovulpes the o2checkcode fails on warnings

**Pull request merged**